File: code/python/dfs_i.py
# global variables
INPUT_PATH = '/home/freyr/dev/GraphTheory/code/python/input.txt'
OUTPUT_PATH = '/home/freyr/dev/GraphTheory/code/python/output.txt'

# clearing the contents of the file if any
open(OUTPUT_PATH, 'w').close()


class Edge:
    def __init__(self, src, dest, cost=1):
        self.src = src
        self.dest = dest
        self.cost = cost
        

class Graph:
    def __init__(self, num_nodes, edges, directed=False):
        self.num_nodes = num_nodes
        self.adj_list = {x : [] for x in range(num_nodes)}
        
        for edge in edges:
            self.adj_list[edge.src].append((edge.dest, edge.cost))
            if not directed:
                self.adj_list[edge.dest].append((edge.src, edge.cost))
                

def dfs(graph, visited, v=0):
    stack = []
    stack.append(v)
    visited[v] = True
    
    while(len(stack) != 0):
        node = stack.pop()
        
        # printing out the node
        with open(OUTPUT_PATH, "a") as f:
            f.write(f"{node} ")
        
        for u in graph.adj_list[node]:
            if not visited[u[0]]:
                stack.append(u[0])
                visited[u[0]] = True

def dfs_count(graph, visited, v=0):
    stack = []
    stack.append(v)
    visited[v] = True
    count = 0
    
    while(len(stack) != 0):
        node = stack.pop()
        count += 1
        
        for u in graph.adj_list[node]:
            if not visited[u[0]]:
                stack.append(u[0])
                visited[u[0]] = True
                
    return count


if __name__ == '__main__':
    
    with open(INPUT_PATH, "r") as f:
        num_nodes, num_edges = (int(i) for i in f.readline().split(" "))
        directed, weighted = (int(i) for i in f.readline().split(" "))
        
        edges = []
        if weighted:
            for _ in range(num_edges):
                u, v, w = (int(i) for i in f.readline().split(" "))
                edges.append(Edge(u, v, w))
        else:
            for _ in range(num_edges):
                u, v = (int(i) for i in f.readline().split(" "))
                edges.append(Edge(u, v))
                
    graph = Graph(num_nodes, edges)
    visited1 = [False for _ in range(num_nodes)]
    visited2 = [False for _ in range(num_nodes)]
    count = 0
    
    # Output Indicator
    with open(OUTPUT_PATH, "a") as f:
        f.write("\nPROGRAM OUTPUT STARTS HERE:\n\n")
        
    for i in range(num_nodes):
        if not visited1[i]:
            dfs(graph, visited1, i)
            
        if not visited2[i]:
            count += dfs_count(graph, visited2, i)
            
    with open(OUTPUT_PATH, "a") as f:
        f.write(f"\nNo. of Nodes: {count}")

File: code/cpp/dfs_i.cpp
#include<iostream>

#include<map>
#include<stack>
#include<string>
#include<vector>
#include<algorithm>


class Edge
{
public:
    int src, dest, cost;

    Edge(int s, int d, int c=1)
    {
        src = s;
        dest = d;
        cost = c;
    }
};

class Graph
{
public:
    int numNodes;
    std::map<int, std::vector<std::pair<int, int>>> adjList;

    Graph(int numNodes, std::vector<Edge> &edges, bool directed=false)
    {
        this->numNodes = numNodes;
        for (Edge &edge: edges)
        {
            adjList[edge.src].push_back({edge.dest, edge.cost});
            if (!directed)
                adjList[edge.dest].push_back({edge.src, edge.cost});
        }
    }
};

void dfs(Graph &graph, std::vector<bool> &visited, int v=0)
{
    std::stack<int> s;    
    s.push(v);
    visited[v] = true;

    while(!s.empty())
    {
        int node = s.top();
        s.pop();
        std::cout << node << " ";
        // if (visited[node]) continue;

        for (auto u : graph.adjList[node])
        {
            if (!visited[u.first])
            {
                s.push(u.first);
                visited[u.first] = true;
            }                
        }
    }
}

// finding connected components in the graph
long dfsCountNode(Graph &graph, std::vector<bool> &visited, int v=0)
{
    std::stack<int> s;
    s.push(v);
    visited[v] = true;
    long count = 0;

    while(!s.empty())
    {
        int node = s.top();
        s.pop();
        count++;
        // if (visited[node]) continue;

        for (auto u : graph.adjList[node])
        {
            if (!visited[u.first])
            {
                s.push(u.first);
                visited[u.first] = true;
            }
        }
    }

    return count;
}


int main()
{
#ifndef FILE_INPUT
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif

    int numNodes, numEdges;
    // std::cout << "Enter no. of Nodes and no. of Edges in the graph ...\n";
    std::cin >> numNodes >> numEdges;

    bool directed, weighted;
    // std::cout << "Is the Graph weighted and directed (0/1) ...\n";
    std::cin >> weighted >> directed;

    int u, v, w;
    std::vector<Edge> edges;
    if (weighted)
    {
        for (int i=0; i<numEdges; i++)
        {
            std::cin >> u >> v >> w;
            edges.push_back(Edge(u, v, w));
        }
    }
    else
    {
        for (int i=0; i<numEdges; i++)
        {
            std::cin >> u >> v;
            edges.push_back(Edge(u, v));
        }
    }

    // Initializing Graph and visited array
    Graph graph = Graph(numNodes, edges, directed);
    std::vector<bool> visited1(numNodes, false);
    std::vector<bool> visited2(numNodes, false);


    long count = 0;

    for (int i=0; i<numNodes; i++)
    {
        if (!visited1[i])
            dfs(graph, visited1, i);
        if (!visited2[i])
            count += dfsCountNode(graph, visited2, i);
    }

    std::cout << "\nNumber of nodes: " << count; 
    return 0;
}